aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -554,15 +816,165 @@ module aptos_framework::delegation_pool {
+    fun update_and_borrow_mut_delegator_vote_delegation(
+        pool: &DelegationPool,
+        governance_records :&mut GovernanceRecords,
+        delegator: address
+    ): &mut VoteDelegation {
+        let (latest_vote_delegation, is_existed, needs_update) =
+            get_latest_delegator_vote_delegation(pool, governance_records, delegator);
+        let vote_delegation_table = &mut governance_records.vote_delegation;
+        if (!is_existed) {
+            smart_table::add(vote_delegation_table, delegator, latest_vote_delegation);
+        };
+        let vote_delegation = smart_table::borrow_mut(vote_delegation_table, delegator);
+        if (needs_update) {
+            *vote_delegation = latest_vote_delegation;
+        };
+        vote_delegation
+    }
+
+    /// Get the latest VoteDelegation of a voter.
+    /// The 2nd bool return value is true when VoteDelegation has already been created before calling this function.
+    /// The 3rd bool return value is true when the latest VoteDelegation is different from the existing VoteDelegation.
+    fun get_latest_vote_delegation(
+        pool: &DelegationPool,
+        governance_records: &GovernanceRecords,
+        voter: address
+    ): (DelegatedVotes, bool, bool) {
+        let pool_address = get_pool_address(pool);
+        let locked_until_secs = stake::get_lockup_secs(pool_address);
+        let voting_power_per_voter = &governance_records.delegated_votes;
+        // By default, a delegator's voter is itself.
+        if (!smart_table::contains(voting_power_per_voter, voter)) {
+            let active_shares = get_delegator_active_shares(pool, voter);
+            let inactive_shares = get_delegator_pending_inactive_shares(pool, voter);
+            return (DelegatedVotes {
+                active_shares,
+                pending_inactive_shares: inactive_shares,
+                active_shares_next_lockup: active_shares,
+                last_locked_until_secs: locked_until_secs,
+            }, false, false)
+        };
+        let latest_vote_delegation = *smart_table::borrow(voting_power_per_voter, voter);
+        let needs_updated = false;
+        if (latest_vote_delegation.last_locked_until_secs < locked_until_secs) {
+            latest_vote_delegation.active_shares = latest_vote_delegation.active_shares_next_lockup;
+            latest_vote_delegation.pending_inactive_shares = 0;
+            latest_vote_delegation.active_shares_next_lockup = latest_vote_delegation.active_shares;
+            latest_vote_delegation.last_locked_until_secs = locked_until_secs;
+            needs_updated = true;
+        };
+        (latest_vote_delegation, true, needs_updated)
+    }
+
+    /// Update VoterVotingPower of a voter to up-to-date then borrow_mut it.
+    fun update_and_borrow_mut_vote_delegation(
+        pool: &DelegationPool,
+        governance_records :&mut GovernanceRecords,
+        voter: address
+    ): &mut DelegatedVotes {
```

Comment:
This get latest and update logic is a bit janky similar to update_and_borrow_mut_delegator_vote_delegation. Anything we can do here to simplify?